extensions/typescript-language-features/src/languageFeatures/diagnostics.ts

```diff
@@ -148,6 +150,76 @@ class DiagnosticSettings {
 	}
 }
 
+class DiagnosticsTelemetryManager extends Disposable {
+
+	private readonly _diagnosticCodesMap = new Map<number, number>();
+	private readonly _diagnosticSnapshotsMap = new Map<vscode.Uri, readonly vscode.Diagnostic[]>();
+	private _timeout: NodeJS.Timeout | undefined;
+
+	constructor(
+		private readonly _telemetryReporter: TelemetryReporter,
+		private readonly _getDiagnostics: (uri: vscode.Uri) => readonly vscode.Diagnostic[]
+	) {
+		super();
+		this._register(vscode.workspace.onDidChangeTextDocument(e => {
+			if (e.document.languageId === 'typescript') {
+				this._updateDiagnosticCodesAfterTimeoutForUri(e.document.uri);
+			}
+		}));
+		this._updateAllDiagnosticCodesAfterTimeout();
+		this._registerTelemetryEventEmitter();
+	}
+
+	private _updateAllDiagnosticCodesAfterTimeout() {
+		const uris = vscode.workspace.textDocuments.map(doc => doc.uri);
+		uris.forEach(uri => setTimeout(() => { this._updateDiagnosticCodes(uri); }, 10000));
+	}
+
+	private _updateDiagnosticCodesAfterTimeoutForUri(uri: vscode.Uri) {
+		clearTimeout(this._timeout);
+		this._timeout = setTimeout(() => { this._updateDiagnosticCodes(uri); }, 10000);
+	}
+
+	private _updateDiagnosticCodes(uri: vscode.Uri) {
+		const previousDiagnostics = this._diagnosticSnapshotsMap.get(uri);
+		const currentDiagnostics = this._getDiagnostics(uri);
+		this._diagnosticSnapshotsMap.set(uri, currentDiagnostics);
+		const diagnosticsDiff = currentDiagnostics.filter((diagnostic) => !previousDiagnostics?.some((previousDiagnostic) => JSON.stringify(diagnostic) === JSON.stringify(previousDiagnostic)));
+		diagnosticsDiff.forEach((diagnostic) => {
+			const code = diagnostic.code;
+			if (typeof code === 'string' || typeof code === 'number') {
+				this._diagnosticCodesMap.set(Number(code), (this._diagnosticCodesMap.get(Number(code)) || 0) + 1);
+			} else if (code !== undefined) {
+				this._diagnosticCodesMap.set(Number(code.value), (this._diagnosticCodesMap.get(Number(code.value)) || 0) + 1);
+			}
```

Comment:
We could extract this to a function like getNumericCode(diagnostic). Then, the map logic wouldn't be duplicated.